GpsDataCapturer/sharedlibrary/src/main/java/com/google/sharedlibrary/GpxFile.java

```diff
@@ -33,54 +36,105 @@ public static File createGpsDataFolder(Context context) {
     }
 
     /**
-     * Create gpxFile if not exist
+     * Create a new gpxFile
+     *
+     * @param gpxDataFolder the folder to store the new file
+     * @param fileName      the name of new file
+     * @return return new created file
      */
     public static File createGpxFile(File gpxDataFolder, String fileName) {
```

Comment:
I like your idea about GpxFile class. Right now it only has static methods.
What if put state inside of it: move 'File gpxFile' and isNew from service to here. Service will be holding refrence to GpsFile instance. And this method createGpsDataFolder is a factory method: creates and returns new GpxFile. Other methods would become instance methods, for example:
public void writeToFile(Context context, Location location) {...}
that is no need to pass gpxFile or isNew as those are member fields now
 - see comments below.
Just an idea.

Reply:
Let me try with that. Another issue is that I couldn't find a way to put the xml file footer into the `GpxFileWriterHandler`. I can set a flag when the service `stopLocationManager/stopFusedLocationProvider `when they _remove the location listener_, but the writing happens _on location changed only_, which means the writer would never get the flag to write footer. 
Another possible solution I could think is that writing the footer with header same time when it's a new file. But then we need to find the writing position when writing captured data. BufferedWriter couldn't do that, instead, it will need  RandomAccessFile to find the writing position. But RandomAccessFile lacks buffering.